### PR TITLE
8065554: MatchResult should provide values of named-capturing groups

### PR DESCRIPTION
**Pull request by rgiulietti**
Add support for named groups to java.util.regex.MatchResult

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires a CSR request to be approved

### Issues
 * [JDK-8065554](https://bugs.openjdk.org/browse/JDK-8065554): MatchResult should provide values of named-capturing groups
 * [JDK-8292872](https://bugs.openjdk.org/browse/JDK-8292872): MatchResult should provide values of named-capturing groups (**CSR**)

### Reviewers
 * [Stuart Marks](https://openjdk.org/census#smarks) (@stuart-marks - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10000/head:pull/10000` \
`$ git checkout pull/10000`

Update a local copy of the PR: \
`$ git checkout pull/10000` \
`$ git pull https://git.openjdk.org/jdk pull/10000/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10000`

View PR using the GUI difftool: \
`$ git pr show -t 10000`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10000.diff">https://git.openjdk.org/jdk/pull/10000.diff</a>

</details>

### CODE REVIEW

File: src/java.base/share/classes/java/util/regex/MatchResult.java [hunk 1]
```diff
@@ -170,7 +237,45 @@ public interface MatchResult {
...
+     *          If there is no capturing group in the pattern
+     *          with the given name
+     *
+     * @implNote The default implementation of this method makes use of the
+     *          map returned by {@link #namedGroups()}.
+     *          It is thus sufficient to override {@link #namedGroups()} for
+     *          this method to work. However, overriding this method directly
+     *          might be preferable for other reasons.
+     *
+     * @since 20
```
[the remark this answers is not in the document]
Comment (rgiulietti) [reply src/java.base/share/classes/java/util/regex/MatchResult.java:274]: Not sure.
If the convention is to document every single `RuntimeException` that methods invoked by this one could throw, then yes.
In other words, should `RuntimeExcpetion`s thrown deep in an invocation stack be documented in every caller method?

[the remark this answers is not in the document]
Comment (dfuch) [reply src/java.base/share/classes/java/util/regex/MatchResult.java:274]: In principle, yes. In practice, I see that `namedGroups` doesn't have an `@throws UnsupportedOperationException` but has an `@implSpec` that says that the default implementation throws `UnsupportedOperationException`. This seems strange to me - maybe @stuart-marks or @jddarcy can comment.

What I was hinting at here however is that we might want to extend the `@implSpec` of the new methods to note that these method will throw `UnsuportedOperationException` if `namedGroups` is not implemented (like the `@implSpec` of `namedGroups` does).

[the remark this answers is not in the document]
Comment (rgiulietti) [reply src/java.base/share/classes/java/util/regex/MatchResult.java:274]: Ah, I see.
So what you mean is not adding another @throws clause but to either improve @implNote or, better, to add @implSpec.

File: src/java.base/share/classes/java/util/regex/MatchResult.java [hunk 2]
```diff
@@ -74,7 +77,44 @@ public interface MatchResult {
...
+     * @implNote
+     * The default implementation of this method makes use of the map returned
+     * by {@link #namedGroups()}. It is thus sufficient to override
+     * {@link #namedGroups()} for this method to work. However, overriding this
+     * method directly might be preferable for performance or for other reasons.
+     *
+     * @implSpec
+     * The default implementation of this method throws
+     * {@link UnsupportedOperationException} if {@link #namedGroups()} is not
+     * overridden.
```
Comment (stuart-marks) [inline src/java.base/share/classes/java/util/regex/MatchResult.java:111]: The essential thing for `@implSpec` is to describe "self-use" of methods on this object. This is important for subclassers to know whether they can inherit the default implementation or whether they should override it. It looks like `start(String)` does the following:

- calls namedGroups() to obtain a mapping from group names to group numbers, propagating UOE if namedGroups() throws it
- if `name` is not present in the group map, throws IAE
- calls `start()` on the group number obtained from the map, and returns that value

I don't think we need to go to the level of detail about whether `get` or `containsKey` is called on the map, but I think the self-calls to `namedGroups()` and `start(int)` are important.

Similar comments apply to the `@implSpec` comments of `end(String)` and `group(String)`.

File: src/java.base/share/classes/java/util/regex/MatchResult.java [hunk 3]
```diff
@@ -74,7 +77,44 @@ public interface MatchResult {
...
+     *
+     * @throws  IllegalArgumentException
+     *          If there is no capturing group in the pattern
+     *          with the given name
+     *
+     * @implNote
+     * The default implementation of this method makes use of the map returned
+     * by {@link #namedGroups()}. It is thus sufficient to override
+     * {@link #namedGroups()} for this method to work. However, overriding this
+     * method directly might be preferable for performance or for other reasons.
```
Comment (stuart-marks) [inline src/java.base/share/classes/java/util/regex/MatchResult.java:106]: This `@implNote` text, is repeated in three different methods. Consider moving this to the class specification. It might make it a bit easier for implementors to see a central overview instead of having this information in each method.

File: src/java.base/share/classes/java/util/regex/MatchResult.java [hunk 4]
```diff
@@ -184,6 +304,53 @@ public interface MatchResult {
...
+     * a previous match or find operation.
+     *
+     * @return whether {@code this} contains a valid match
+     *
+     * @implSpec
+     * The default implementation of this method throws
+     * {@link UnsupportedOperationException}.
+     *
+     * @apiNote
+     * This method must be overridden by an implementation.
```
Comment (stuart-marks) [inline src/java.base/share/classes/java/util/regex/MatchResult.java:348]: This is a bit odd. It sounds like existing MatchResult implementations (outside the JDK) are now invalid. I think it really means something like, "This method must be overridden by an implementation in order to provide valid information about whether this MatchResult contains a match." I'm not sure whether saying this is necessary; it could be omitted.

Probably also needs an `@throws UnsupportedOperationException` in case the match information is unavailable.

File: test/jdk/java/util/regex/NamedGroupsTests.java [hunk 5]
```diff
@@ -0,0 +1,277 @@
...
+    private static void testMatchResultStartEndGroup3() {
+        Pattern.compile("(?<some>.+?)(?<rest>.*)").matcher("")
+                .results()
+                .forEach(r -> {
+                        try {
+                            r.start("noSuchGroup");
+                            r.end("noSuchGroup");
+                            r.group("noSuchGroup");
+                        } catch (IllegalArgumentException e) {  // swallowing intended
+                        }
```
Comment (stuart-marks) [inline test/jdk/java/util/regex/NamedGroupsTests.java:115]: If MatchResult is behind correctly, the call to `r.start("noSuchGroup")` will always throw an exception and the subsequent calls to `r.end` and `r.group` will never be executed. This potentially will miss testing of those methods.

File: test/jdk/java/util/regex/NamedGroupsTests.java [hunk 6]
```diff
@@ -0,0 +1,277 @@
...
+    private static void testMatchResultStartEndGroup4() {
+        List<MatchResult> list = Pattern.compile("(?<some>.+?)(?<rest>.*)").matcher("abc")
+                .results().toList();
+        for (var result : list) {
+            try {
+                result.start("noSuchGroup");
+                result.end("noSuchGroup");
+                result.group("noSuchGroup");
+            } catch (IllegalArgumentException e) {  // swallowing intended
+            }
```
Comment (stuart-marks) [inline test/jdk/java/util/regex/NamedGroupsTests.java:128]: Similar issue here as above.

File: src/java.base/share/classes/java/util/regex/MatchResult.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (stuart-marks) [reply src/java.base/share/classes/java/util/regex/MatchResult.java:289]: I left some inline comments on the `@implSpec`. But I do think that these methods require `@throws UnsupportedOperationException` for the cases where they don't support named groups.

**Comment by stuart-marks** [comment]
Overall the specs, code, and tests look pretty good. I do think some areas of the spec need updating; sorry I didn't get to this before you created the CSR.

The test is OK, but it's starting to get to the point where it would be profitable to use TEST-NG data providers to collapse some of the test cases. We have two implementations of MatchResult: one is Matcher itself, and the other is the internal implementation returned by toMatchResult(). The setup for them differs somewhat, but the assertions should all be the same. This is kind of hard to see with separate test methods for Matcher and MatchResult. The test is reasonable as it stands, but we'll see what it looks like after the cases for checking exceptions from start(String), end(String), and group(String) are expanded.

**Comment by rgiulietti** [comment]
Addressed concerns about spec details.
CSR will be updated once the discussion about the spec has settled and the wording has stabilized.

**Comment by stuart-marks** [comment]
Good updates to the test, the `@throws` clauses, and the `@implNote` in the class specification. I don't know if I was clear before, but all the new default methods require `@implSpec` clauses that explain in some detail what they do, in particular self-use -- see my [previous comment](https://github.com/openjdk/jdk/pull/10000#discussion_r955530412). (Or maybe you're still working on this.)

**Comment by rgiulietti** [comment]
If the most recent commit is OK in terms of Javadoc, I'll update the CSR accordingly

**Comment by stuart-marks** [comment]
Spec looks good. Let me know when you're done updating the CSR so I can mark it reviewed.

File: test/jdk/java/util/regex/NamedGroupsTests.java [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+        testMatchResultHasMatch();
+
+        testMatchResultStartEndGroup();
+    }
+
+    private static void testMatchResultStartEndGroup() {
+        testMatchResultStartEndGroup1();
+        testMatchResultStartEndGroup2();
+        testMatchResultStartEndGroup3();
+    }
```
Comment (stuart-marks) [inline test/jdk/java/util/regex/NamedGroupsTests.java:51]: The three numbered tests here are a little hard to follow. Looks like these tests are

1. test existing group names, with no match
2. test existing group names, with a successful match
3. test nonexistent group names, with a successful match

On test names, sometimes people provide extremely verbose test names such as `testThatExistingGroupNameWithMatchReturnsNegativeOrNull` which I think is overkill, but having a name that's somewhat descriptive would be helpful.

It looks like a case is missing, which is a test for a nonexistent group name on a MatchResult that doesn't have a successful match. I'm not sure which is checked first; I think the implementation would throw IAE, because of the nonexistent name, regardless of whether or not the MatchResult has a match.

However, I don't think we've specified this, and in fact I don't think we want to. In general though, if multiple error conditions can arise in the same operation, the general style is not to constrain implementations to check for things in a particular order. Thus either IAE or ISE would be acceptable. Perhaps a test should be added for that. (Hm, might want to take another look at the specs regarding this issue.)

File: test/jdk/java/util/regex/NamedGroupsTests.java [hunk 8]
```diff
@@ -0,0 +1,267 @@
...
+    public static void main(String[] args) {
+        testPatternNamedGroups();
+        testMatcherNamedGroups();
+        testMatchResultNamedGroups();
+
+        testMatcherHasMatch();
+        testMatchResultHasMatch();
+
+        testMatchResultStartEndGroup();
+    }
```
Comment (stuart-marks) [inline test/jdk/java/util/regex/NamedGroupsTests.java:45]: I haven't gone through all the tests in great detail (yet), but it occurs to me that we potentially have THREE implementations of some of the logic, and strictly speaking we should test all the code paths. The three implementations are in:

1. Matcher
2. Matcher$ImmutableMatchResult
3. MatchResult's default methods

I took a quick look and it looks like Matcher and Matcher$ImmutableMatchResult override the default methods, so the default methods themselves need to be tested. This is essentially testing the `@implSpec`. The typical way to do that is to have the test create its own MatchResult implementation(s). There might need to be implementations that do and do not implement `namedGroups`, in order to test UOE. They might also need some state to cover various cases of no-match, has-match with and without group names, etc.

File: test/jdk/java/util/regex/NamedGroupsTests.java (hunk 7 above)
Comment (rgiulietti) [reply test/jdk/java/util/regex/NamedGroupsTests.java:51]: > (Hm, might want to take another look at the specs regarding this issue.)

Not sure who wants to take another look. If that it's you, then I'll wait with the CSR.
I'll change the method names to something a bit more speaking.

File: test/jdk/java/util/regex/NamedGroupsTests.java (hunk 8 above)
Comment (rgiulietti) [reply test/jdk/java/util/regex/NamedGroupsTests.java:45]: An implementation that overrides `namedGroups` without overriding the other methods accepting group names is `Matcher$ImmutableMatchResult`, which is already exercised in the tests.

File: test/jdk/java/util/regex/NamedGroupsTests.java (hunk 7 above)
Comment (stuart-marks) [reply test/jdk/java/util/regex/NamedGroupsTests.java:51]: Sorry I should have been more specific. "Somebody" should take another look. :-) Well, anyway, I did, and the specification as written does not indicate which error condition is checked first. I think this is OK, so I don't think any changes are necessary. You might mention this in the text of the CSR; I know that Joe and I have discussed this issue previously, and he might have a recommendation.

File: test/jdk/java/util/regex/NamedGroupsTests.java (hunk 8 above)
Comment (stuart-marks) [reply test/jdk/java/util/regex/NamedGroupsTests.java:45]: OK, as long as all the code paths are covered.

**Comment by rgiulietti** [comment]
Added an implementation of `MatchResult` that does not override any of the default methods.
More speaking method names.

**Comment by rgiulietti** [comment]
CSR updated to current status.

File: src/java.base/share/classes/java/util/regex/MatchResult.java [hunk 9]
```diff
@@ -184,6 +307,49 @@ public interface MatchResult {
...
+    int groupCount();
+
+    /**
+     * Returns an unmodifiable map from capturing group names to group numbers.
+     * If there are no named groups, returns an empty map.
+     *
+     * @return an unmodifiable map from capturing group names to group numbers
+     *
+     * @throws UnsupportedOperationException
+     *          The default implementation of this method always throws
```
Comment (stuart-marks) [inline src/java.base/share/classes/java/util/regex/MatchResult.java:319]: The `@throws` clause needs to state a general contract over all implementations, so it should say something like, "@throws UOE if the implementation does not support named groups". Then, the specification for the default implementation always throwing UOE should be moved to `@implSpec`.

File: src/java.base/share/classes/java/util/regex/MatchResult.java [hunk 10]
```diff
@@ -184,6 +307,49 @@ public interface MatchResult {
...
+    }
+
+    /**
+     * Returns whether {@code this} contains a valid match from
+     * a previous match or find operation.
+     *
+     * @return whether {@code this} contains a valid match
+     *
+     * @throws UnsupportedOperationException
+     *          The default implementation of this method always throws
```
Comment (stuart-marks) [inline src/java.base/share/classes/java/util/regex/MatchResult.java:347]: Similar comment here as above, though it has nothing to do with named groups. The `@throws` clause should say it throws UOE if "the implementation cannot report whether or not it has a match" or some such. This is a bit odd, but the specification needs to be permissive enough so that it doesn't invalidate existing implementations outside the JDK.

As before, the "always throws" should be moved to `@implSpec`.

File: src/java.base/share/classes/java/util/regex/MatchResult.java (hunk 9 above)
Comment (rgiulietti) [reply src/java.base/share/classes/java/util/regex/MatchResult.java]: Done

File: src/java.base/share/classes/java/util/regex/MatchResult.java (hunk 10 above)
Comment (rgiulietti) [reply src/java.base/share/classes/java/util/regex/MatchResult.java]: Done

**Review by stuart-marks — APPROVED** [review]
OK I took another look at everything and I think this is good to integrate.

The tests seem adequate but it seems like they would benefit from some refactoring. It might be an interesting exercise to revisit them and try out the new JUnit 5 APIs.

**Comment by rgiulietti** [comment]
I'll integrate now but agree to open another PR to make use of JUnit in the tests, perhaps later this week or next week.
